src/python/pants/engine/internals/specs_rules.py

```diff
@@ -255,35 +261,63 @@ async def resolve_specs_snapshot(
 ) -> SpecsSnapshot:
     """Resolve all files matching the given specs.
 
-    Address specs will use their `SourcesField` field, and Filesystem specs will use whatever args
-    were given. Filesystem specs may safely refer to files with no owning target.
+    All matched targets will use their `sources` field. Certain specs like FileLiteralSpec will
+    also match against all their files, regardless of if a target owns them.
+
+    If a file is owned by a target that gets filtered out (e.g. via `--tag`), then we make sure
+    the file is not added back via filesystem specs, per
+    https://github.com/pantsbuild/pants/issues/15478.
     """
-    targets = await Get(Targets, SpecsWithoutFileOwners, SpecsWithoutFileOwners.from_specs(specs))
+
+    unfiltered_targets = await Get(
+        Targets, Specs, dataclasses.replace(specs, filter_by_global_options=False)
+    )
+    filtered_targets = await Get(FilteredTargets, Targets, unfiltered_targets)
     all_hydrated_sources = await MultiGet(
         Get(HydratedSources, HydrateSourcesRequest(tgt[SourcesField]))
-        for tgt in targets
+        for tgt in filtered_targets
         if tgt.has_field(SourcesField)
     )
 
+    digests = [hydrated_sources.snapshot.digest for hydrated_sources in all_hydrated_sources]
+
     with_files_owners = SpecsWithOnlyFileOwners.from_specs(specs)
-    filesystem_specs_digest = (
-        await Get(
+    filtered_out_sources_paths: set[str] = set()
+    if with_files_owners:
```

Comment:
There are two consecutive `if with_files_owners` blocks... if that's right, maybe merge them or comment to explain why?

Reply:
Oh good eye.